### ISSUE TITLE
Roslyn compiles C#14 extension members with incorrect nullability constraints metadata

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#79896]
**Version Used**: 2e23a0df5bf8d49f6bf856287b0eb10de08f97d9

**Steps to Reproduce**:

Compile the following code:
```csharp
#nullable enable
#pragma warning disable CS8631

I? iNull = null;
E.M(iNull); // CS8714 is reported if the library is compiled, no warnings if referenced as source code
iNull.M(); // always no warnings
```

Which references this library:
```csharp
#nullable enable

public interface I { }

public static class E
{
    extension<T>(T t) where T : I
    {
        public T M() => t;
    }
}
```

**Expected Behavior**:
Nullability warnings are the same regardless of whether you place the extension in the same project or in a compiled assembly.

**Actual Behavior**:
The following warning is reported when the extension comes from a compiled assembly but not if you use it from source code:

```warning CS8714: The type 'I?' cannot be used as type parameter 'T' in the generic type or method 'E.M<T>(T)'. Nullability of type argument 'I?' doesn't match 'notnull' constraint.```

**Notes**:
The behavior is caused by a bogus `notnull` constraint encoded in the metadata for the extension method compiled for `E.M`. The constraint doesn't exist in the source code and therefore is not accounted for when the extension code is used from sources.

Note that `E.M` extension is present in the metadata twice, as a static member of `E` and as a member of the nested extension type. These two copies have different type parameter constraints. The nested extension member doesn't include the bogus `notnull` constraint, which is why the extension form of the call doesn't depend on whether you are using a compiled assembly or not.

E.M:
```
  .method public hidebysig static !!0/*T*/
    M<(class I) T>(
      !!0/*T*/ t
    ) cil managed
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.ExtensionAttribute::.ctor()
      = (01 00 00 00 )
    .param [0]
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(unsigned int8)
        = (01 00 01 00 00 ) // .....
        // unsigned int8(1) // 0x01
    .param constraint [1] /*T*/, class I
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(unsigned int8)
        = (01 00 01 00 00 ) // .....
        // unsigned int8(1) // 0x01
```

E.<>E__0`1. extension:
```
  .class nested public sealed auto ansi beforefieldinit
    '<>E__0`1'<(class I) T>
      extends [System.Runtime]System.Object
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableContextAttribute::.ctor(unsigned int8)
      = (01 00 01 00 00 ) // .....
      // unsigned int8(1) // 0x01
    .param type [1] /*T*/
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(unsigned int8)
        = (01 00 00 00 00 ) // .....
        // unsigned int8(0) // 0x00
```

<details>
<summary>Full compiled IL:</summary>

```
.class public abstract sealed auto ansi beforefieldinit
  E
    extends [System.Runtime]System.Object
{
  .custom instance void [System.Runtime]System.Runtime.CompilerServices.ExtensionAttribute::.ctor()
    = (01 00 00 00 )

  .class nested public sealed auto ansi beforefieldinit
    '<>E__0`1'<(class I) T>
      extends [System.Runtime]System.Object
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableContextAttribute::.ctor(unsigned int8)
      = (01 00 01 00 00 ) // .....
      // unsigned int8(1) // 0x01
    .param type [1] /*T*/
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(unsigned int8)
        = (01 00 00 00 00 ) // .....
        // unsigned int8(0) // 0x00

    .method private hidebysig static specialname void
      '<Extension>$'(
        !0/*T*/ t
      ) cil managed
    {
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor()
        = (01 00 00 00 )
      .maxstack 8

      IL_0000: ret

    } // end of method '<>E__0`1'::'<Extension>$'

    .method public hidebysig instance !0/*T*/
      M() cil managed
    {
      .maxstack 8

      IL_0000: ldnull
      IL_0001: throw

    } // end of method '<>E__0`1'::M
  } // end of class '<>E__0`1'

  .method public hidebysig static !!0/*T*/
    M<(class I) T>(
      !!0/*T*/ t
    ) cil managed
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.ExtensionAttribute::.ctor()
      = (01 00 00 00 )
    .param [0]
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(unsigned int8)
        = (01 00 01 00 00 ) // .....
        // unsigned int8(1) // 0x01
    .param constraint [1] /*T*/, class I
      .custom instance void [System.Runtime]System.Runtime.CompilerServices.NullableAttribute::.ctor(unsigned int8)
        = (01 00 01 00 00 ) // .....
        // unsigned int8(1) // 0x01
    .maxstack 8

    // [9 25 - 9 26]
    IL_0000: ldarg.0      // t
    IL_0001: ret

  } // end of method E::M
} // end of class E
```
</details>

<details>
<summary>Roslyn test code to verify:</summary>

Neither empty expected diagnostics nor CS8714 pass both verify calls, showing the difference between compiled and source-code behavior
```csharp
var src = """
          #nullable enable
          #pragma warning disable CS8631

          I? iNull = null;
          E.M(iNull);
          iNull.M();
          """;
var libSrc = """
             #nullable enable

             public interface I { }

             public static class E
             {
                 extension<T>(T t) where T : I
                 {
                     public T M() => t;
                 }
             }
             """;
DiagnosticDescription[] expected = [
    // (5,1): warning CS8714: The type 'I?' cannot be used as type parameter 'T' in the generic type or method 'E.M<T>(T)'. Nullability of type argument 'I?' doesn't match 'notnull' constraint.
    // E.M(iNull);
    Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterNotNullConstraint, "E.M").WithArguments("E.M<T>(T)", "T", "I?").WithLocation(5, 1)
];

var comp = CreateCompilation([src, libSrc]);
comp.VerifyEmitDiagnostics(expected);

var libComp = CreateCompilation(libSrc);
var comp2 = CreateCompilation(src, references: [libComp.EmitToImageReference()]);
comp2.VerifyEmitDiagnostics(expected);
```
</details>

**Comment by colejohnson66** [comment]
This is still an issue on 10.0.20x.

**Comment by CyrusNajmabadi** [comment]
@colejohnson66 The issue is still open. :)

**Comment by RikkiGibson** [comment]
This is weird. `I` is a non-nullable type, so, constraining `T` to it means that converting `null` to `T` should always produce a warning.
@jcouv am I remembering something wrong? I can see that the analogous classic extension scenario also doesn't warn. The body assumes the parameter is non-null but the call site doesn't enforce a non-null is flowing in..

[.NET Lab](https://lab.razor.fyi/#vZHBihNBEIZZwUVaBsQnKNjLDEgvg6Ji3A2SDcvIjgiZkyc7PZVJYU937OohG5bAvoX4Al48eN-H2IfZJ5CeKMSIOXpqur7_r_-HEt8OhXjvXeNVKzU__nLYMdkGJisO2A4Ssf2VI2cM6kDOsjxHi570PknRtl1QU4O7ojNSjXUcSPMeJEeuxjdWmRXTX7oLsp93Z9Xco6rJNv8EslL8Ke5KxJHtjInlAG18EnG08KppFSyVt9FcE_d8NHn5_GkePcUQ6F1nDJxAdA8SMZZl2o-yARwfR-mL_BkQg8eF8wFroBmEOYKhqVd-FZF27YIM1k_Aut9hHHUeZ-jRaqxBMbDrvEbQrsZE9BmyTDcxyizVirftsd1YlnmaFsPM_rc--d5Ci25qSAPZgH6mNEIBV7DeIhxUIA3aKGYYJ-IqEQAAf1IOnmwzhDJ_XZ2mYU4MFYQMlnP0CBW8ggJOTiHIyk16aZr1B46b8DKgZXI2WnddG8mvzK3cCso061cONmydiPXbe5o_PHxwe3N3_eP71-Gj-x8PLg9-Ag)

If you want to allow null to flow in you can constrain `T : I?` instead.

**Comment by TessenR** [comment]
@RikkiGibson Base type nullability constraint mismatch is reported as CS8631 
```
warning CS8631: The type 'I?' cannot be used as type parameter 'T' in the generic type or method 'E.M<T>(T)'. Nullability of type argument 'I?' doesn't match constraint type 'I'.
```

It's suppressed in this snippet because it is always correctly emitted.

To get CS8714 you have to additionally specify a redundant `notnull` constraint on top of the base type being non-nullable, like this:
```
#nullable enable
public static class E
{
    public static string? M1<T>(this T t) where T : I => t.ToString();

    extension<T>(T t) where T : notnull, I
    {
        public T M() => t;
    }
}
```

or, put the extension in a compiled assembly and the bogus non-existent constraint will be added to the metadata

**Comment by RikkiGibson** [comment]
Ah, I missed what that pragma was doing, thanks.

So, all the usages in the sample produce warnings, it's just, a different warning or set of warnings might be reported depending on where the extension is declared.

It sounds like all we want to change here is to stop placing a notnull constraint on the metadata version of the type parameter.

**Comment by colejohnson66** [comment]
Yes. For our use case, we had the following:

```csharp
public static class NotifyPropertyChangedExtensions
{
  public static IObservable<TValue> WhenAnyValue<TTarget, TValue>(
    this TTarget target,
    string name,
    Func<TTarget, TValue> getter)
    where TTarget : INotifyPropertyChanged;
}
```

Usage with null values was acceptable (e.g., `this.WhenAnyValue(nameof(SelectedX), static x => x.SelectedX);`);

When upgrading to extension syntax as below, Roslyn started putting `where TTarget:notnull` and `where TValue:notnull`.

```csharp
public static class NotifyPropertyChangedExtensions
{
  extension<TTarget>(TTarget target)
    where TTarget : INotifyPropertyChanged
  {
    public IObservable<TValue> WhenAnyValue<TValue>(
      string name,
      Func<TTarget, TValue> getter);
  }
}
```

---

There's a `[NullableContext(1)]` (default is `notnull`?) across the new syntax form.

```csharp
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Security.Permissions;

[assembly: Extension]
[assembly: CompilationRelaxations(8)]
[assembly: RuntimeCompatibility(WrapNonExceptionThrows = true)]
[assembly: Debuggable(DebuggableAttribute.DebuggingModes.Default | DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints | DebuggableAttribute.DebuggingModes.EnableEditAndContinue | DebuggableAttribute.DebuggingModes.DisableOptimizations)]
[assembly: SecurityPermission(SecurityAction.RequestMinimum, SkipVerification = true)]
[assembly: AssemblyVersion("0.0.0.0")]
[module: UnverifiableCode]
[module: RefSafetyRules(11)]
[Extension]
public static class NotifyPropertyChangedExtensions
{
  [NullableContext(1)]
  [Extension]
  public static IObservable<TValue> WhenAnyValue<[Nullable(0)] TTarget, [Nullable(2)] TValue>(TTarget target, string name, Func<TTarget, TValue> getter) where TTarget : INotifyPropertyChanged
  {
    throw new NotImplementedException();
  }
}
[Extension]
public static class NotifyPropertyChangedExtensions2
{
  [SpecialName]
  [Extension]
  public sealed class <G>$B87FB99B0FC93086A396186AFB24677B<$T0> where $T0 : INotifyPropertyChanged
  {
    [SpecialName]
    public static class <M>$3CBC9DE089886EB156A1FBD7635A9CD0
    {
      [SpecialName]
      [NullableContext(1)]
      [CompilerGenerated]
      public static void <Extension>$($T0 target)
      {
      }
    }

    [NullableContext(1)]
    [ExtensionMarker("<M>$3CBC9DE089886EB156A1FBD7635A9CD0")]
    public IObservable<TValue> WhenAnyValue<[Nullable(2)] TValue>(string name, Func<$T0, TValue> getter)
    {
      throw new NotSupportedException();
    }
  }

  [NullableContext(1)]
  [Extension]
  [return: Nullable(new byte[] { 1, 0 })]
  public static IObservable<TValue> WhenAnyValue<TTarget, TValue>(TTarget target, string name, [Nullable(new byte[] { 1, 1, 0 })] Func<TTarget, TValue> getter) where TTarget : INotifyPropertyChanged
  {
    throw new NotImplementedException();
  }
}
```
